core/src/main/scala/kafka/network/SocketServer.scala

```diff
@@ -492,6 +453,71 @@ private[kafka] class Processor(val id: Int,
     }
   }
 
+  /* `protected` for test usage */
+  protected[network] def sendResponse(response: RequestChannel.Response) {
+    trace(s"Socket server received response to send, registering for write and sending data: $response")
+    val channel = selector.channel(response.responseSend.destination)
+    // `channel` can be null if the selector closed the connection because it was idle for too long
```

Comment:
Yes, another possibility is that the client closes the socket after request timeout. The server can detect the cancelled key even if the READ interest bit is off. Then, when the server tries to send a response, the socket channel no longer exists. In the typical path, it seems the way the cancelled key is detected is that the key will be marked ready for reads, then when you read it, it returns a -1 size. However, in theory in may be possible for the selector to detect an invalid key directly w/o going through read?